src/coreclr/jit/promotiondecomposition.cpp

```diff
@@ -24,13 +25,1181 @@ class DecompositionStatementList
+            statements->AddStatement(m_compiler->gtNewAssignNode(dst, src));
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // CreateInitValue:
+    //   Create an IR node representing a constant value with the specified init pattern.
+    //
+    // Parameters:
+    //   type        - The primitive type
+    //   initPattern - Pattern to init with
+    //
+    // Returns:
+    //   A constant.
+    //
+    // Remarks:
+    //   Should only be called when that pattern can actually be represented;
+    //   for example, SIMD types and GC pointers only support an init pattern
+    //   of zero.
+    //
+    GenTree* CreateInitValue(var_types type, int64_t initPattern)
+    {
+        switch (type)
+        {
+            case TYP_BOOL:
+            case TYP_BYTE:
+            case TYP_UBYTE:
+            case TYP_SHORT:
+            case TYP_USHORT:
+            case TYP_INT:
+            {
+                int64_t mask = (int64_t(1) << (genTypeSize(type) * 8)) - 1;
+                return m_compiler->gtNewIconNode(static_cast<int32_t>(initPattern & mask));
+            }
+            case TYP_LONG:
+                return m_compiler->gtNewLconNode(initPattern);
+            case TYP_FLOAT:
+                float floatPattern;
+                memcpy(&floatPattern, &initPattern, sizeof(floatPattern));
+                return m_compiler->gtNewDconNode(floatPattern, TYP_FLOAT);
+            case TYP_DOUBLE:
+                double doublePattern;
+                memcpy(&doublePattern, &initPattern, sizeof(doublePattern));
+                return m_compiler->gtNewDconNode(doublePattern);
+            case TYP_REF:
+            case TYP_BYREF:
+#ifdef FEATURE_SIMD
+            case TYP_SIMD8:
+            case TYP_SIMD12:
+            case TYP_SIMD16:
+#if defined(TARGET_XARCH)
+            case TYP_SIMD32:
+            case TYP_SIMD64:
+#endif // TARGET_XARCH
+#endif // FEATURE_SIMD
+            {
+                assert(initPattern == 0);
+                return m_compiler->gtNewZeroConNode(type);
+            }
+            default:
+                unreached();
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // FinalizeCopy:
+    //   Create IR to perform the decomposed copy.
+    //
+    // Parameters:
+    //   statements - List to add statements to.
+    //
+    void FinalizeCopy(DecompositionStatementList* statements)
+    {
+        assert(m_dst->OperIs(GT_LCL_VAR, GT_LCL_FLD, GT_BLK) && m_src->OperIs(GT_LCL_VAR, GT_LCL_FLD, GT_BLK));
+
+        RemainderStrategy remainderStrategy = DetermineRemainderStrategy();
+
+        // If the remainder is a full block and is going to incur write barrier
+        // then avoid incurring multiple write barriers for each source
+        // replacement that is a GC pointer -- write them back to the struct
+        // first instead.
```

Comment:
Will add a comment on this specific optimization. For completeness:
```cpp
        // If the remainder is a full block and is going to incur write barrier
        // then avoid incurring multiple write barriers for each source
        // replacement that is a GC pointer -- write them back to the struct
        // first instead. That is, instead of:
        //
        //   ▌  COMMA     void  
        //   ├──▌  ASG       struct (copy)                      <- write barrier
        //   │  ├──▌  BLK       struct<Program+S, 32>
        //   │  │  └──▌  LCL_VAR   byref  V01 arg1         
        //   │  └──▌  LCL_VAR   struct<Program+S, 32> V00 arg0         
        //   └──▌  COMMA     void  
        //      ├──▌  ASG       ref                             <- write barrier
        //      │  ├──▌  IND       ref   
        //      │  │  └──▌  ADD       byref 
        //      │  │     ├──▌  LCL_VAR   byref  V01 arg1         
        //      │  │     └──▌  CNS_INT   long   8
        //      │  └──▌  LCL_VAR   ref    V05 tmp3         
        //      └──▌  ASG       ref                             <- write barrier
        //         ├──▌  IND       ref   
        //         │  └──▌  ADD       byref 
        //         │     ├──▌  LCL_VAR   byref  V01 arg1         
        //         │     └──▌  CNS_INT   long   24
        //         └──▌  LCL_VAR   ref    V06 tmp4         
        //
        // Produce:
        // 
        //   ▌  COMMA     void  
        //   ├──▌  ASG       ref                                <- no write barrier
        //   │  ├──▌  LCL_FLD   ref    V00 arg0         [+8]
        //   │  └──▌  LCL_VAR   ref    V05 tmp3         
        //   └──▌  COMMA     void  
        //      ├──▌  ASG       ref                             <- no write barrier
        //      │  ├──▌  LCL_FLD   ref    V00 arg0         [+24] 
        //      │  └──▌  LCL_VAR   ref    V06 tmp4         
        //      └──▌  ASG       struct (copy)                   <- write barrier
        //         ├──▌  BLK       struct<Program+S, 32>
        //         │  └──▌  LCL_VAR   byref  V01 arg1          (last use)
        //         └──▌  LCL_VAR   struct<Program+S, 32> V00 arg0         
        //
```